course_discovery/apps/edx_haystack_extensions/elasticsearch_boost_config.py

```diff
@@ -156,7 +156,7 @@ def get_elasticsearch_boost_config():
                             }
                         }
                     },
-                    'weight': -20.0
+                    'weight': 0.0
```

Comment:
Why this change?

Reply:
I needed to leave a comment here, which I have already done.

The answer is simple - es7 does not support negative values for function_score boosting.
Here are links that talk about it and suggest how you can go out and find another solution.

https://github.com/elastic/elasticsearch/issues/24910
https://github.com/elastic/elasticsearch/issues/31927

Here is the answer from es7 when trying to use negative values.
`elasticsearch.exceptions.RequestError: RequestError(400, 'illegal_argument_exception', '[weight] cannot be negative for a filtering function')`

Now, this filter makes no sense at all and does not affect the overall score in any way. This is a problem. We need to think about what to do with it now. Perhaps it should be abandoned and processed differently. Here I want to consult with you. What do you think?

Reply:
I'm not very knowledgeable about boosting, but could we raise this boost to 0.0, then bump all the other boosts in this file up 20? Would that result in the same relative boosts? 

It also seems like everything else in this config file already has a weight above zero. So does the magnitude difference really matter? Is zero just as good as -20 here? (this is my ignorance of boosting showing)

Reply:
Let's share with you how boosts work.
Let's say we make a request to elasticsearch server and 10 documents will perfectly satisfy our request query. All 10 documents have a score of `1.0`. All documents are in a row. But we are not satisfied with their order and we know the criteria by which we would be able to display the results more successfully, so that the most necessary documents, in our opinion, were at the top. This is where boosts come in. For example, there is a boost
```
{
                    'filter': {
                        'term': {
                            'pacing_type': 'self_paced'
                        }
                    },
                    'weight': 5.0
                },
```
and two of the 10 documents have `pacing_type` equal to` self_paced` since we have `'boost_mode ':' sum '` and `` score_mode ':' sum'` then for these two documents the score will now be `1.0 + 5.0 = 6.0` so these two documents will be above all. And so on. I think the main idea is clear.
If `'weight': 0.0`, then `1.0 + 0.0 = 1.0`, i.e. such a boost is meaningless and it does not affect the score of the documents in any way. The idea here was that the documents for which boost is below, we need to greatly lower the score and so that they are displayed as low as possible in the list of found results.
```
{
                    'filter': {
                        'bool': {
                            'must': {
                                'term': {
                                    'content_type': 'courserun'
                                }
                            },
                            'must_not': {
                                'range': {
                                    'paid_seat_enrollment_end': {
                                        'gte': 'now'
                                    }
                                }
                            }
                        }
                    },
                    'weight': 0.0
                }
```
This is how it all works. Now I don't know what to do. The easiest option is to refuse this boost or look for an alternative. In those links that I put there may be answers.
But unfortunately, leaving `'weight': 0.0` is pointless, and adding 20 to all is also useless since it will not affect the score of the documents.

Reply:
If it only matters (with our config) that the resulting score is relatively larger or smaller, zero seems as good as -20, in a world where nothing else is at zero. I guess that's the question -- are there documents that don't have a matching config in this file that are also getting a default zero weight? Probably... :-/

Can we invert the logic of the filter, and boost everything *that wouldn't match* by 20? i.e. boost "course runs *with* enrollable, paid seats" by 20

Reply:
> If it only matters (with our config) that the resulting score is relatively larger or smaller

It's true. At any time, any who knows what to do cat manipulate it whatever they want.

Reply:
> Can we invert the logic of the filter, and boost everything that wouldn't match by 20? i.e. boost "course runs with enrollable, paid seats" by 20

Michale - you are good! That's why I wrote that I need your ideas. And sometimes the fresh view from aside is really that you are looking for. Yes, I think we can do this and it's a good workaround.

Reply:
I modified the boost and here is it - https://github.com/raccoongang/course-discovery/pull/12/commits/55e776830f1ef44a435fbc8d8905ebb786cf3d0e
The goal maintains and works, but in a different manner, i.e. vise verse.